### PR TITLE
NCSDK-5199: AT parser not handling all modem responses correctly

### PR DESCRIPTION
**Pull request by dagbja**
Temporary fix for failing commands (HWVERSION etc.)

Signed-off-by: Dag Bjarvin <Dag.Bjarvin@nordicsemi.no>

### CODE REVIEW

**Comment by nordic-hani** [comment]
I think the correct prefix for the PR would be "lib:at_cmd_parser: "

File: lib/at_cmd_parser/at_cmd_parser.c [hunk 1]
```diff
@@ -251,6 +268,8 @@ static int at_parse_param(const char **at_params_str,
 
 	reset_state();
 
+	setTypeString = false;
```
Comment (nordic-hani) [inline lib/at_cmd_parser/at_cmd_parser.c:271]: Could ``setTypeString = false;`` simply be inside of the ``reset_state()`` function?

File: lib/at_cmd_parser/at_cmd_parser.c [hunk 2]
```diff
@@ -61,6 +63,21 @@ static int at_parse_detect_type(const char **str, int index)
 		 * notification ID, (eg +CEREG:)
 		 */
 		set_new_state(NOTIFICATION);
+
+		if (!strncmp(tmpstr, "%HWVERSION", 10)) {
```
Comment (nordic-hani) [inline lib/at_cmd_parser/at_cmd_parser.c:67]: Should these numbers be defines? ``AT_CMD_HWVERSION_LEN`` ```AT_CMD_SHORTSWVER_LEN`` etc?

File: lib/at_cmd_parser/at_cmd_parser.c [hunk 3]
```diff
@@ -61,6 +63,21 @@ static int at_parse_detect_type(const char **str, int index)
...
+				setTypeString = true;
+		}
+		if (!strncmp(tmpstr, "%SHORTSWVER", 11)) {
```
Comment (nordic-hani) [inline lib/at_cmd_parser/at_cmd_parser.c:70]: Would it be more efficient for the following ifs be ``esle if``? Looks like they can not occur at the same time.

**Comment by nordic-hani** [comment]
This section where we check that the notification is in fact not a string: maybe it could be extracted out to it's own small function? Might be easier to explain and understand what. And to hopefully remove one day.

File: lib/at_cmd_parser/at_cmd_parser.c [hunk 4]
```diff
@@ -31,6 +36,8 @@ enum at_parser_state {
 
 static enum at_parser_state state;
 
+static bool setTypeString = false;
```
Comment (rlubos) [inline lib/at_cmd_parser/at_cmd_parser.c:39]: snake_case please (`set_type_string`).

Comment (dagbja) [reply lib/at_cmd_parser/at_cmd_parser.c:39]: Fixed

File: lib/at_cmd_parser/at_cmd_parser.c (hunk 1 above)
Comment (dagbja) [reply lib/at_cmd_parser/at_cmd_parser.c]: Done

File: lib/at_cmd_parser/at_cmd_parser.c (hunk 2 above)
Comment (dagbja) [reply lib/at_cmd_parser/at_cmd_parser.c]: Fixed

File: lib/at_cmd_parser/at_cmd_parser.c (hunk 3 above)
Comment (dagbja) [reply lib/at_cmd_parser/at_cmd_parser.c]: Done

File: lib/at_cmd_parser/at_cmd_parser.c [hunk 5]
```diff
@@ -31,6 +36,8 @@ enum at_parser_state {
 
 static enum at_parser_state state;
 
+static bool set_type_string = false;
```
Comment (rlubos) [inline lib/at_cmd_parser/at_cmd_parser.c:39]: Checkpatch complains about initializing static bool with false (which will be implicitly initialized), lets make it (`static bool set_type_string`).

File: lib/at_cmd_parser/at_cmd_parser.c [hunk 6]
```diff
@@ -52,6 +61,26 @@ static inline void skip_command_prefix(const char **cmd)
 	(*cmd)++;
 }
 
+static inline bool check_response_for_forced_string(const char *tmpstr)
+{
+	bool retval = false;
+
+	if (!strncmp(tmpstr, "%HWVERSION", AT_CMD_HWVERSION_LEN)) {
+			retval = true;
```
Comment (rlubos) [inline lib/at_cmd_parser/at_cmd_parser.c:69]: The indentation is too deep, should be one tab less.

File: lib/at_cmd_parser/at_cmd_parser.c [hunk 7]
```diff
@@ -52,6 +61,26 @@ static inline void skip_command_prefix(const char **cmd)
...
+	}
+	else if (!strncmp(tmpstr, "%SHORTSWVER", AT_CMD_SHORTSWVER_LEN)) {
```
Comment (rlubos) [inline lib/at_cmd_parser/at_cmd_parser.c:71]: This should be at the same line as closing bracket, i. e. `} else if (...) {`.

Or personally I'd even consider making this whole expression shorter:
```
if (!strncmp(tmpstr, "%HWVERSION", AT_CMD_HWVERSION_LEN) ||
    !strncmp(tmpstr, "%SHORTSWVER", AT_CMD_SHORTSWVER_LEN) ||
    !strncmp(tmpstr, "%XMODEMUUID", AT_CMD_XMODEMUUID_LEN) ||
    !strncmp(tmpstr, "%XICCID", AT_CMD_XICCID_LEN)) {
	retval = true;
}
```

**Review by rlubos — COMMENTED** [review]
Pointed out `checkpatch` complaints.

There's also a space missing after `lib: ` in the commit title, `gitlint` reports it.

File: lib/at_cmd_parser/at_cmd_parser.c (hunk 7 above)
Comment (dagbja) [reply lib/at_cmd_parser/at_cmd_parser.c]: Agreed. Fixed.

File: lib/at_cmd_parser/at_cmd_parser.c (hunk 6 above)
Comment (dagbja) [reply lib/at_cmd_parser/at_cmd_parser.c]: Fixed.

File: lib/at_cmd_parser/at_cmd_parser.c (hunk 5 above)
Comment (dagbja) [reply lib/at_cmd_parser/at_cmd_parser.c]: Agreed. Fixed.

**Comment by rlubos** [comment]
Something went wrong with the PR branch, no clue why there are so many merge commits (there shouldn't be any).

File: lib/at_cmd_parser/at_cmd_parser.c [hunk 8]
```diff
@@ -52,6 +65,23 @@ static inline void skip_command_prefix(const char **cmd)
 	(*cmd)++;
 }
 
+static inline bool check_response_for_forced_string(const char *tmpstr)
+{
+	bool retval = false;
+
+	if (!strncmp(tmpstr, "+CGSN", AT_CMD_CGSN_LEN) ||
```
Comment (evenl) [inline lib/at_cmd_parser/at_cmd_parser.c:72]: This don't make sense for AT+CGSN.

AT+CGSN will only return the serial number without any notification identifier (+CGSN: ). For AT+CGSN=1, AT+CGSN=2, and AT+CGSN=3 the values are already returned with quotes. So they will be parsed as strings anyway.

Modem firmware version 1.2.1-57.rc

**Comment by dagbja** [comment]
The commit history got really messed up. I'll start a new PR and will delete this one - probably easier than fixing this.

**Pull request closed without merging**